### PR TITLE
Add stochastic subgrid example

### PR DESCRIPTION
**Pull request by jlippuner**
## PR Summary

Add an example that performs dummy FLOP work for a random number of iterations in each cell. The number of iterations is sampled from a power-law distribution whose coefficient and minimum and maximum numbers of iterations can be specified in the input file.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [n/a] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md

### CODE REVIEW

**Comment by jlippuner** [comment]
@Yurlungur @jdolence This is ready for you to take a look and see if the dummy work is adequate for our purposes (see function `DoLotsOfWork` on line 338 of `advanced_advection_package.cpp`. Thanks!

File: example/advanced_advection/CMakeLists.txt
Comment (Yurlungur) [inline example/advanced_advection/CMakeLists.txt]: I don't love the name `advanced_advection-example`. Could we come up with something slightly more descriptive? Like `heterogeneous-advection-example` or something?

File: example/advanced_advection/advanced_advection_driver.cpp
Comment (Yurlungur) [inline example/advanced_advection/advanced_advection_driver.cpp]: I'm not sure what to do about this if logic here. It's in the simple advection example. But for the more complicated one, I vote we pick one... if you're using sparse variables, I say no packs. If you're not using sparse variables, I saw do the packs.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [inline example/advanced_advection/advanced_advection_package.cpp]: I'm not sure this actually works... should lambda not be the diagonal accross the domain? I.e., it should be a pythagorean rule?

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [inline example/advanced_advection/advanced_advection_package.cpp]: Why do you hand implement this instead of using the stdlib? Just curious.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [inline example/advanced_advection/advanced_advection_package.cpp]: Why the mdrange loop pattern tag?

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [inline example/advanced_advection/advanced_advection_package.cpp]: I'm not sure if this matters, but it might be good to change these exponentials to `pow(10,x)` and the logarithms to `log10`. Riot uses base 10... and if I recall, that's slower.

Actually to be most precise, riot uses `logf(x)*ILOG10` where `constexpr Real ilog10 = 1./log(10);`

File: src/pgen/default_pgen.cpp
Comment (Yurlungur) [inline src/pgen/default_pgen.cpp:92]: good catch regarding these

**Review by Yurlungur — APPROVED** [review]
Thanks for this, @jlippuner ! Overall, this looks great! I have a few comments/questions/suggestions, but they're all non-blocking.

**Comment by Yurlungur** [comment]
@jlippuner some profiling information for FillDerrived in RIOT. The following should be considered a lower bound on cost, but I think it's reasonable. The average number of iterations (in the test problem I ran) is about 4 with a long tail.
- On CPU, we have about 1 solve/ms.
- On GPU, we have about 133 solves/ms.

Note these times are over all cells. e.g., I took total time and divide by total number of cells.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (jlippuner) [reply example/advanced_advection/advanced_advection_package.cpp]: Good question. I just copied this from the `advection` example. Maybe @pgrete can shed some light on this since it appears he wrote this code in the `advection` example.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (jlippuner) [reply example/advanced_advection/advanced_advection_package.cpp]: Excellent question. I'm always for using stdlib and not reinventing (or re-implementing) the wheel. I did look at `std::discrete_distribution` and considered using it, but eventually decided to implement the alias method myself for several reasons:

1. It's not clear to me which method `std::discrete_distribution` uses internally. It says it's `operator()` method is amortized constant, so it could be less efficient than the alias method, which is guaranteed constant.
2. It's also not clear to me if `std::discrete_distribution` keeps any dynamic internal state and thus if it is thread-safe, which we would need, otherwise we'd have to make many copies of it, one for each thread.
3. Most importantly, since we sample the number of iterations on the device, we would need `std::discrete_distribution` to work on the GPU in parallel. I didn't try it and it's possible that it works, but I decided I'd rather know exactly what's happening on the GPU and how much memory we use. With the alias method implemented ourselves, we just read from two small `Kokkos::view`s (the probability and alias tables) of which we only need one copy and the actual sampling just involves two table look ups without any dynamic state, so we know it's thread-safe (given that Kokkos' random number generator is thread-safe) and lightweight on the device.

It may be interesting to try to use `std::discrete_distribution` on the GPU, but I suspect we will get some additional UVM transfers because we can't tell `std::discrete_distribution` to allocate its internal state on the device, as we can with `Kokkos::view`s.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (jlippuner) [reply example/advanced_advection/advanced_advection_package.cpp]: No particular reason. That's just what I saw in other examples. Should I use something else?

File: example/advanced_advection/advanced_advection_package.cpp
Comment (pgrete) [reply example/advanced_advection/advanced_advection_package.cpp]: I recycled the code from the Athena++ pgen so unless I made a copy/paste error this should work.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (jlippuner) [reply example/advanced_advection/advanced_advection_package.cpp]: I'm happy to change to base 10. For the log, should I use `log10` or multiply by `1/log(10)`? Where do you get the `constexpr` version of the `log` function to define the constant `ilog10`? When I try to do this, it complains that `cannot call non-constexpr function "log(double)"`

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [reply example/advanced_advection/advanced_advection_package.cpp]: Ok. I must not fully understand what's going on here. If it works in advection and Athena++ it's probably fine.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [reply example/advanced_advection/advanced_advection_package.cpp]: I see. Interesting.

Yeah I don't think `std::discrete_distribution` would work on GPU. I think the strategy there would be to call it on CPU and precompute random numbers, and then fill a kokkos view with them.

Would it be worth saving this alias method to a `parthenon` utility so it's available in the future and to downstream codes?

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [reply example/advanced_advection/advanced_advection_package.cpp]: If it runs without specifying a loop pattern, we probably want to not specify one and let the default tag do its work. But that might depend on the dimension of the loop.... I think only only `mdrange` is supported for 5D and 6D loops. (Which is something we should get around to fixing at some point.)

Comment (Yurlungur) [reply example/advanced_advection/advanced_advection_package.cpp]: Sorry I forgot about that detail until just now. I guess let's leave this for now. We can fix it when we add support for 5d and 6d loops with the other tags.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (Yurlungur) [reply example/advanced_advection/advanced_advection_package.cpp]: I think I computed `1/log(10)` and then copy-pasted the number to get around the lack of a constexpr for log. Let's just multiply by `1/log(10)`. That probably is sufficient.

File: example/advanced_advection/advanced_advection_package.cpp
Comment (jlippuner) [reply example/advanced_advection/advanced_advection_package.cpp]: I think that's a good idea. I'll split it off.

File: example/advanced_advection/CMakeLists.txt
Comment (jlippuner) [reply example/advanced_advection/CMakeLists.txt]: Thanks @Yurlungur for the suggestion `stochastic_subgrid`!

**Comment by jlippuner** [comment]
This is good to go once CI passes. I need one more review. @pgrete @JoshuaSBrown @cielling ?

File: example/stochastic_subgrid/parthenon_app_inputs.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/parthenon_app_inputs.cpp]: Can be made const.

File: example/stochastic_subgrid/parthenon_app_inputs.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/parthenon_app_inputs.cpp]: Can be made const.

File: example/stochastic_subgrid/parthenon_app_inputs.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/parthenon_app_inputs.cpp]: const reference?

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_package.cpp:45]: This is a large function I would recommend breaking this up into smaller local functions in an unnames namespace. I think a lot of the variables in this function can be declared const, that would help clarify where things are being changed vs being used.

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_package.cpp:217]: Can `v` be made const reference?

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_package.cpp]: make const.

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_package.cpp]: make const

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_package.cpp]: const

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_package.cpp]: const

File: example/stochastic_subgrid/parthinput.stochastic_subgrid
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/parthinput.stochastic_subgrid:54]: What is cfl? Could we use something a little more descriptive?

File: example/stochastic_subgrid/stochastic_subgrid_driver.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_driver.cpp]: There is a lot of repetition here, I would recommend placing in a local function.

File: example/stochastic_subgrid/stochastic_subgrid_driver.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_driver.cpp:59]: This method is really long, I would recommend breaking into small local functions.

Comment (JoshuaSBrown) [reply example/stochastic_subgrid/stochastic_subgrid_driver.cpp:59]: I think the method can also be made const, unless I'm missing something.

File: src/utils/alias_method.cpp
Comment (JoshuaSBrown) [inline src/utils/alias_method.cpp]: Can you use std accumulate?

File: src/utils/alias_method.cpp
Comment (JoshuaSBrown) [inline src/utils/alias_method.cpp]: This is more stylistic but I prefer putting opening and closing braces to prevent accidentally shifting lines and changing logic.

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_package.cpp]: Make const

**Review by JoshuaSBrown — CHANGES_REQUESTED** [review]
It looks good to me, it is very ordered and clean pr. I would like to see a few of the larger functions broken up into smaller sub functions to make the logic a little clearer. There was a little repetition as well which would benefit from the same. Everything else I think is just small stuff. Does it make sense to add a regression test for this example or is it basically covering the same logic as the advection examples? Actually, I think this would make a really good performance tests sense you are testing the performance when the workload is unevenly distributed. If we can pick the random seeds it could be really useful for testing the stability of the performance.

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (jlippuner) [reply example/stochastic_subgrid/stochastic_subgrid_package.cpp:45]: I don't think breaking it up makes sense, because there is no code reuse here that could be extracted. So breaking up this function into smaller functions is the same as breaking it up into scopes with comment headers that delineate different sections. That's what I've done now. I think that's a better way to go because it doesn't unnecessarily bloat the code with functions that are called only once.

I've rearranged the reading and adding parameters to the package to keep things as local as possible and I've also added const to the variables that are not changed.

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (jlippuner) [reply example/stochastic_subgrid/stochastic_subgrid_package.cpp:217]: Hmm, good question. I think it could, but I don't think we gain anything. It goes into a `KOKKOS_LAMBDA` that captures it by value and thus makes a copy and I think the Kokkos documentation encourages to use `view` by value since it's very lightweight and can do reference counting properly that way.

File: example/stochastic_subgrid/parthinput.stochastic_subgrid
Comment (jlippuner) [reply example/stochastic_subgrid/parthinput.stochastic_subgrid:54]: CFL is the Courant-Friedrichs-Lewy condition. This is extremely common and widely used in hydrodynamics and computational fluid dynamics. Although I'm usually always for more descriptive variable names, because "cfl" is used so ubiquitously, I think it would actually be more confusing to name it something different, because people expect it to be called "cfl".

Comment (JoshuaSBrown) [reply example/stochastic_subgrid/parthinput.stochastic_subgrid:54]: Ok, I guess my ignorance of this knowledge domain is showing. :)

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (JoshuaSBrown) [reply example/stochastic_subgrid/stochastic_subgrid_package.cpp:45]: Fair enough this is on the side of stylistic preference, you do gain one more advantage by placing things in functions, you can fit more into your screen.

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (jlippuner) [reply example/stochastic_subgrid/stochastic_subgrid_package.cpp]: Hah, I thought you missed this one :P

File: example/stochastic_subgrid/parthinput.stochastic_subgrid
Comment (jlippuner) [reply example/stochastic_subgrid/parthinput.stochastic_subgrid:54]: It's a fair question, though. I'm sure there are some variables that could be more descriptive.

File: example/stochastic_subgrid/stochastic_subgrid_package.cpp
Comment (jlippuner) [reply example/stochastic_subgrid/stochastic_subgrid_package.cpp:45]: Actually, scopes are even better in this regard, I think. You don't have to waste lines pulling in variables you need from the outer scope and you can place the "function body" right where you need it (so no jumping around), and you can fold it just the same to make space for more code on the screen :)

File: example/stochastic_subgrid/stochastic_subgrid_driver.cpp
Comment (jlippuner) [reply example/stochastic_subgrid/stochastic_subgrid_driver.cpp:59]: As per the previous discussion, I prefer to not break up this function into smaller functions because there would be no re-use of those functions. Instead, I added some scoping with header comments to break up the function into manageable sections.

This particular method could be made `const` indeed, however, it's overriding a virtual method in the base class, which is not `const`. I'm not sure if the method in the base class should be const or not. There may be a good reason that we don't want it to be const so that other derived classes have the freedom to do non-`const` things in these methods.

File: src/utils/alias_method.cpp
Comment (jlippuner) [reply src/utils/alias_method.cpp]: Good idea!

File: example/stochastic_subgrid/stochastic_subgrid_driver.cpp
Comment (JoshuaSBrown) [reply example/stochastic_subgrid/stochastic_subgrid_driver.cpp:59]: There is a workaround, add a const method with the same name to the base class, and have the non constant method in your specefic case call the const method. 

https://riptutorial.com/cplusplus/example/16974/avoiding-duplication-of-code-in-const-and-non-const-getter-methods-

This is non-blocking, but I think it would make it easier to follow in the future, if people are not changing the state.

File: src/utils/alias_method.cpp
Comment (jlippuner) [reply src/utils/alias_method.cpp]: Even better:
```cxx
(host_prob_table(i) < 1.0 ? under_full : over_full).push(i);
```

Comment (JoshuaSBrown) [reply src/utils/alias_method.cpp]: under_full is missing a push but ya that is clear. 

Oh, nevermind.

**Comment by jlippuner** [comment]
@JoshuaSBrown Thanks for all your diligent comments. I think I've implemented or responded to all of them. Please review those changes.

File: example/stochastic_subgrid/stochastic_subgrid_driver.cpp
Comment (JoshuaSBrown) [inline example/stochastic_subgrid/stochastic_subgrid_driver.cpp:157]: Yes, this is nicer, it clearly limits the scope of local variables.

**Review by JoshuaSBrown — APPROVED** [review]
Beautiful pr, one last point pleas update the copyright date on all the files you have changed some of them are out of date. Oh, and you never answered me about creating a performance regression test?

**Comment by jlippuner** [comment]
> Beautiful pr, one last point pleas update the copyright date on all the files you have changed some of them are out of date. Oh, and you never answered me about creating a performance regression test?

Thanks and good catch! I'll update that and then merge once CI passes.

**Comment by JoshuaSBrown** [comment]
> > Beautiful pr, one last point pleas update the copyright date on all the files you have changed some of them are out of date. Oh, and you never answered me about creating a performance regression test?
> 
> Thanks and good catch! I'll update that and then merge once CI passes.

**and you never answered me about creating a performance regression test?**

**Comment by jlippuner** [comment]
> > > Beautiful pr, one last point pleas update the copyright date on all the files you have changed some of them are out of date. Oh, and you never answered me about creating a performance regression test?
> > 
> > 
> > Thanks and good catch! I'll update that and then merge once CI passes.
> 
> **and you never answered me about creating a performance regression test?**

Oh yes, sorry I forgot about that. That might be worthwhile. Although I'm not sure we want two performance regression tests. Maybe we should use this one instead of the advection example. I'm not sure. I think that would be something to discuss with the team. I'd stay it's beyond the scope of this PR, though.

**Pull request merged**
